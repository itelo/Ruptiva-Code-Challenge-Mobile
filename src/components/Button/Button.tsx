import React from "react";
import { TouchableOpacity, View, GestureResponderEvent, Text } from "react-native";
import {
  PulseIndicator,
} from 'react-native-indicators';
import { ThemeType } from "@utils/styles/ThemeContext";
import useStyles from "@utils/styles/useStyles";

type ButtonProps = {
  labelTestID?: string;
  touchableTestID?: string;
  disabled?: boolean;
  isLoading?: boolean;
  label: string;
  loadingLabel: string;
  onPress(event: GestureResponderEvent): void;
}

const Button = (props: ButtonProps) => {
  const styles = useStyles(classes)
  return (
    <TouchableOpacity
      testID={props.touchableTestID}
      disabled={props.disabled}
      style={styles.buttonRoot}
      onPress={props.onPress}
    >
      {props.isLoading ? <View style={styles.fixButtonContentWhenLoading}>
        <View>
          <PulseIndicator color='white' />
        </View>
        <Text testID={props.labelTestID} style={styles.buttonLabel}>{props.loadingLabel}</Text>
      </View> :
        <Text testID={props.labelTestID} style={styles.buttonLabel}>{props.label}</Text>}
    </TouchableOpacity>
  )
}

Button.defaultProps = {
  disabled: false,
  isLoading: false
};

const classes = (theme: ThemeType) => ({
  buttonRoot: {
    marginHorizontal: 48,
    paddingVertical: 12,
    paddingHorizontal: 12,
    height: 48,
    alignItems: "center" as "center",
    justifyContent: "center" as "center",
    flexDirection: "row" as "row",
    borderTopRightRadius: 0,
    borderTopLeftRadius: 0,
    backgroundColor: theme.palette.boring.blue
  },
  buttonLabel: {
    ...theme.typography.button,
    color: "#FFF"
  },
  fixButtonContentWhenLoading: {
    marginLeft: -20,
    alignItems: "center" as "center",
    justifyContent: "center" as "center",
    flexDirection: "row" as "row"
  }
});

export default Button;