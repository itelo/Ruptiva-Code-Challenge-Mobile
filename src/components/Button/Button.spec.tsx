import React from "react";
import Button from "./Button";
import { render, fireEvent } from "react-native-testing-library";
import renderer from "react-test-renderer";

describe("<Button />", () => {
  it("fire onPress event", () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <Button
        touchableTestID="button"
        onPress={onPressMock}
        label="button"
        loadingLabel="loading"
      />
    );

    const button = getByTestId("button");

    fireEvent.press(button);
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it("get the label name", () => {
    const onPressMock = jest.fn();
    const { getByTestId, debug } = render(
      <Button
        labelTestID="label"
        onPress={onPressMock}
        label="An Awesome Label Here"
        loadingLabel="loading"
      />
    );

    const label = getByTestId("label");
    expect(label.props.children).toBe("An Awesome Label Here");
  });

  it("get the loading label", () => {
    const onPressMock = jest.fn();
    const { getByTestId, debug } = render(
      <Button
        labelTestID="label"
        isLoading={true}
        onPress={onPressMock}
        label="button"
        loadingLabel="An Awesome Label Here"
      />
    );

    const label = getByTestId("label");
    expect(label.props.children).toBe("An Awesome Label Here");
  });

  it("test props isLoading={false} label='button'", () => {
    const onPressMock = jest.fn();
    const { toJSON } = render(
      <Button
        onPress={onPressMock}
        label="button"
        loadingLabel="loading"
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("test props isLoading={true} loadingLabel='loading'", () => {
    const onPressMock = jest.fn();
    const { toJSON } = render(
      <Button
        onPress={onPressMock}
        isLoading={true}
        label="button"
        loadingLabel="loading"
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});