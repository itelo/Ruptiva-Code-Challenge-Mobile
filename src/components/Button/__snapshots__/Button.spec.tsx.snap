// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> test props isLoading={false} label='button' 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b2ff",
      "borderTopLeftRadius": 0,
      "borderTopRightRadius": 0,
      "flexDirection": "row",
      "height": 48,
      "justifyContent": "center",
      "marginHorizontal": 48,
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFF",
        "fontFamily": "SF Pro Text",
        "fontSize": 14,
        "fontWeight": "500",
        "letterSpacing": 0.45712,
        "lineHeight": 24.5,
      }
    }
  >
    button
  </Text>
</View>
`;

exports[`<Button /> test props isLoading={true} loadingLabel='loading' 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b2ff",
      "borderTopLeftRadius": 0,
      "borderTopRightRadius": 0,
      "flexDirection": "row",
      "height": 48,
      "justifyContent": "center",
      "marginHorizontal": 48,
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 12,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginLeft": -20,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          animating={true}
          animationDuration={1200}
          animationEasing={[Function]}
          color="white"
          interaction={true}
          renderComponent={[Function]}
          style={
            Object {
              "height": 40,
              "width": 40,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 20,
                  "height": 40,
                  "opacity": 0.5,
                  "transform": Array [
                    Object {
                      "scale": 0.4,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 20,
                  "height": 40,
                  "opacity": 1,
                  "transform": Array [
                    Object {
                      "scale": 0.4,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "SF Pro Text",
          "fontSize": 14,
          "fontWeight": "500",
          "letterSpacing": 0.45712,
          "lineHeight": 24.5,
        }
      }
    >
      loading
    </Text>
  </View>
</View>
`;
